ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -333,9 +390,14 @@ void addToIndex(IndexEntry entry, FileSpan fileSpan) throws StoreException {
       IndexSegment info = new IndexSegment(dataDir, entry.getValue().getOffset(), factory, entry.getKey().sizeInBytes(),
           entry.getValue().getBytes().capacity(), config, metrics);
       info.addEntry(entry, fileSpan.getEndOffset());
-      indexes.put(info.getStartOffset(), info);
+      indexesChangeLock.lock();
```

Comment:
From what I see, `changeIndexSegments()` will never get called to alter the trailing part of the Index (covered by the `Journal`), and it has explicit checks to guard against that. For the same reason that you do not have the `journal.addEntry()` covered by this lock here, I think the index put doesn't need a lock either. You should be able to deference the atomic reference and simply do the put. Am I missing something?

Also, I don't see the need for the lock in general (see my other comment), but maybe you have other reasons for that which are not part of this patch.